File: examples/javascript/07_bulk_export.js
const BASE = process.env.JETNET_BASE_URL || "https://customer.jetnetconnect.com";
const EMAIL = process.env.JETNET_EMAIL;
const PASS = process.env.JETNET_PASSWORD;

if (!EMAIL || !PASS) {
  console.error("Set JETNET_EMAIL and JETNET_PASSWORD environment variables.");
  process.exit(1);
}

async function login() {
  const res = await fetch(`${BASE}/api/Admin/APILogin`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ emailAddress: EMAIL, password: PASS }),
  });
  if (!res.ok) throw new Error(`Login failed: HTTP ${res.status}`);
  const data = await res.json();
  return {
    bearerToken: data.bearerToken,
    apiToken: data.apiToken || data.securityToken,
  };
}

function isJetnetError(json) {
  return (
    json?.responsestatus &&
    String(json.responsestatus).toUpperCase().includes("ERROR")
  );
}

async function jetnetFetch(bearerToken, apiToken, path, opts = {}) {
  const url = `${BASE}${path}`.replaceAll("{apiToken}", apiToken);
  const res = await fetch(url, {
    ...opts,
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${bearerToken}`,
      ...(opts.headers || {}),
    },
  });
  const json = await res.json();
  if (!res.ok || isJetnetError(json)) {
    throw new Error(
      isJetnetError(json) ? json.responsestatus : `HTTP ${res.status}`
    );
  }
  return json;
}

async function bulkExportAllPages(bearerToken, apiToken, body, pageSize = 100) {
  const allAircraft = [];
  let page = 1;

  while (true) {
    const data = await jetnetFetch(
      bearerToken,
      apiToken,
      `/api/Aircraft/getBulkAircraftExportPaged/{apiToken}/${pageSize}/${page}`,
      { method: "POST", body: JSON.stringify(body) }
    );

    const aircraft = data.aircraft || [];
    allAircraft.push(...aircraft);

    const totalPages = Math.max(data.maxpages ?? 1, 1);
    console.log(
      `  Page ${page} of ${totalPages} — ${aircraft.length} records this page`
    );

    if (page >= totalPages) break;
    page++;
  }

  return allAircraft;
}

async function main() {
  const { bearerToken, apiToken } = await login();

  const body = {
    airframetype: "None",
    maketype: "BusinessJet",
    modlist: [145],
    lifecycle: "InOperation",
    actiondate: "",
    enddate: "",
    forsale: "",
    aclist: [],
    exactMatchReg: false,
    showHistoricalAcRefs: false,
    showAwaitingDocsCompanies: false,
    showMaintenance: false,
    showAdditionalEquip: false,
    showExterior: false,
    showInterior: false,
  };

  console.log("Bulk exporting G550 aircraft (modelid 145)...\n");

  const aircraft = await bulkExportAllPages(bearerToken, apiToken, body);

  console.log(`\nTotal aircraft exported: ${aircraft.length}\n`);

  for (const ac of aircraft.slice(0, 10)) {
    const forSale = ac.forsale === "Y" ? "For Sale" : "";
    const owner = ac.owrcompanyname || "N/A";
    const operator = ac.oprcompanyname || "N/A";

    console.log(
      `  ${ac.regnbr || "N/A"} | ${ac.make} ${ac.model} | SN: ${ac.serialnbr || "N/A"} | Owner: ${owner} | Operator: ${operator} ${forSale}`
    );
  }

  if (aircraft.length > 10) {
    console.log(`\n  ... and ${aircraft.length - 10} more aircraft.`);
  }
}

main().catch((err) => {
  console.error("Error:", err.message);
  process.exit(1);
});

File: examples/python/07_bulk_export.py
"""
JETNET API -- Bulk Aircraft Export with Pagination

Demonstrates how to use getBulkAircraftExportPaged to export aircraft records
in bulk, including the critical maxpages:0 quirk handling.

Key points:
- getBulkAircraftExport returns maxpages:0 when all results fit in one page
- Use max(maxpages, 1) to avoid skipping single-page results
- Response uses flat prefixed schema: owr* = owner, opr* = operator, chp* = chief pilot
- forsale is "Y"/"N" (not "true"/"false")
- datepurchased is YYYYMMDD format

Required environment variables:
    JETNET_EMAIL    -- Your JETNET account email
    JETNET_PASSWORD -- Your JETNET account password
"""

import os
import sys
import requests
from datetime import datetime, timedelta

BASE = "https://customer.jetnetconnect.com"


def login(email, password):
    r = requests.post(
        f"{BASE}/api/Admin/APILogin",
        json={"emailAddress": email, "password": password},
    )
    r.raise_for_status()
    data = r.json()
    return data["bearerToken"], data["apiToken"]


def api(method, path, bearer, token, body=None):
    url = f"{BASE}{path}".replace("{apiToken}", token)
    headers = {
        "Authorization": f"Bearer {bearer}",
        "Content-Type": "application/json",
    }
    r = requests.request(method, url, headers=headers, json=body)
    r.raise_for_status()
    result = r.json()
    if "ERROR" in str(result.get("responsestatus", "")).upper():
        raise ValueError(f"JETNET error: {result['responsestatus']}")
    return result


def get_bulk_export(bearer, token, body, pagesize=100):
    """Fetch all pages of a bulk aircraft export.

    Handles the maxpages:0 quirk where JETNET returns maxpages=0
    when all results fit in a single response.
    """
    all_aircraft = []
    page = 1

    while True:
        data = api(
            "POST",
            f"/api/Aircraft/getBulkAircraftExportPaged/{{apiToken}}/{pagesize}/{page}",
            bearer, token, body,
        )

        aircraft = data.get("aircraft", [])
        all_aircraft.extend(aircraft)

        total_pages = max(data.get("maxpages", 1), 1)

        print(f"  Page {page}/{total_pages} -- {len(aircraft)} records (total so far: {len(all_aircraft)})")

        if page >= total_pages:
            break
        page += 1

    return all_aircraft


def main():
    email = os.environ.get("JETNET_EMAIL")
    password = os.environ.get("JETNET_PASSWORD")
    if not email or not password:
        print("Set JETNET_EMAIL and JETNET_PASSWORD environment variables.")
        sys.exit(1)

    bearer, token = login(email, password)

    print("--- Bulk Export: In-Operation Business Jets by Model ---\n")

    body = {
        "airframetype": "FixedWing",
        "maketype": "BusinessJet",
        "lifecycle": "InOperation",
        "modlist": [145],
        "aclist": [],
        "actiondate": "",
        "forsale": "",
        "exactMatchReg": False,
        "showHistoricalAcRefs": False,
        "showAwaitingDocsCompanies": False,
        "showMaintenance": False,
        "showAdditionalEquip": False,
        "showExterior": False,
        "showInterior": False,
    }

    aircraft = get_bulk_export(bearer, token, body, pagesize=100)
    print(f"\nTotal aircraft exported: {len(aircraft)}\n")

    for i, ac in enumerate(aircraft[:5], 1):
        reg = ac.get("regnbr", "N/A")
        make = ac.get("make", "")
        model = ac.get("model", "")
        sn = ac.get("serialnbr", "N/A")
        year = ac.get("yearmfr", "N/A")
        forsale = ac.get("forsale", "N/A")
        owner = ac.get("owrcompanyname", "N/A")
        operator = ac.get("oprcompanyname", "N/A")

        print(f"  {i}. {reg:10s}  {make} {model}  SN: {sn}  Year: {year}")
        print(f"     For Sale: {forsale}  Owner: {owner}  Operator: {operator}")

    print("\n\n--- Hourly Change Detection (actiondate window) ---\n")

    now = datetime.now()
    one_hour_ago = now - timedelta(hours=1)

    body_hourly = {
        "airframetype": "None",
        "maketype": "BusinessJet",
        "lifecycle": "InOperation",
        "actiondate": one_hour_ago.strftime("%m/%d/%Y %H:%M:%S"),
        "enddate": now.strftime("%m/%d/%Y %H:%M:%S"),
        "exactMatchReg": False,
        "showHistoricalAcRefs": False,
        "showAwaitingDocsCompanies": False,
        "showMaintenance": False,
        "showAdditionalEquip": False,
        "showExterior": False,
        "showInterior": False,
    }

    changes = get_bulk_export(bearer, token, body_hourly, pagesize=100)
    print(f"\nAircraft records changed in the last hour: {len(changes)}")

    for ac in changes[:5]:
        reg = ac.get("regnbr", "N/A")
        make = ac.get("make", "")
        model = ac.get("model", "")
        print(f"  - {reg}  {make} {model}")


if __name__ == "__main__":
    main()
